File: demoNote/src/test/java/ThinkingInJava/InitializeAndClear/Practice.java
package ThinkingInJava.InitializeAndClear;

/**
 * @author Rex
 * @title: Practice
 * @projectName demoNote
 * @description: TODO
 * @date 2019/6/317:04
 */
public class Practice {

    public static void main(String[] args) {

        TestFinalize testFinalize = new TestFinalize(true);
        testFinalize = null;
//        testFinalize.checkIn();

//        new TestFinalize(true);
        System.gc();
    }

}

class TestFinalize{

    boolean checkOut = false;
    TestFinalize(boolean checkOut){
        this.checkOut = checkOut;
    }
    void checkIn(){
        this.checkOut = false;
    }
    protected void finalize(){
        if(checkOut)
        System.out.println("say something");
    }

}

File: scala/src/main/scala/com/atguigu/chapter7/Practice.scala
package com.atguigu.chapter7

/**
  * @title: Practice
  * @projectName scala
  * @description: TODO
  * @author Rex
  * @date 2019/9/2316:19
  */
object Practice {

  def main(args: Array[String]): Unit = {
    val ints: List[Int] = List(1,2,3,4)
    println(listSum(ints))
  }

  /**
    * 计算一个list[Int]的和
    */
  def listSum(list:List[Int]):Int = {
    if(list.size < 1){
      return 0
    }

    if (list.size == 1) {
      return list(0)
    }

    val tmp:Int = list(0) + list(1)
    val ints: List[Int] = tmp+:list.tail.tail
    listSum(ints)
  }

}
